### ISSUE TITLE
Very long waiting time before training starts (15 mins)

### ISSUE THREAD

**Issue by yxchng** [facebookresearch/maskrcnn-benchmark#651]
## 🐛 Bug

Very long waiting time before training starts (15 mins). I am not sure if it is a bug but it feels like one. If not, what do I need to do to speed it up? 

> ...
> 2019-04-09 13:58:35,545 maskrcnn_benchmark.data.build WARNING: When using more than one image per GPU you may encounter an out-of-memory (OOM) error if your GPU does not have sufficient memory. If this happens, you can reduce SOLVER.IMS_PER_BATCH (for training) or TEST.IMS_PER_BATCH (for inference). For training, you must also adjust the learning rate and schedule length according to the linear scaling rule. See for example: https://github.com/facebookresearch/Detectron/blob/master/configs/getting_started/tutorial_1gpu_e2e_faster_rcnn_R-50-FPN.yaml#L14
> 2019-04-09 14:13:06,296 maskrcnn_benchmark.trainer INFO: Start training

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux): 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6.7
 - CUDA/cuDNN version: 10/7.4
 - GPU models and configuration: 1070 Ti
 - Any other relevant information:

**Comment by yxchng** [comment]
This is due to the fact that

```
def _compute_aspect_ratios(dataset):
    aspect_ratios = []
    for i in range(len(dataset)):
        img_info = dataset.get_img_info(i)
        aspect_ratio = float(img_info["height"]) / float(img_info["width"])
        aspect_ratios.append(aspect_ratio)
    return aspect_ratios
```

will call the function `get_img_info` for all images before training starts to `_compute_aspect_ratios`.

Therefore, your implementation for `get_img_info` has to be efficient or pre-computed. My first implementation was inefficient as I read the requested image on the fly, which takes really long as it is required for all images.

The following recommendation on how to create your own dataset should mention this to prevent people from making this mistake.

```
from maskrcnn_benchmark.structures.bounding_box import BoxList

class MyDataset(object):
    def __init__(self, ...):
        # as you would do normally

    def __getitem__(self, idx):
        # load the image as a PIL Image
        image = ...

        # load the bounding boxes as a list of list of boxes
        # in this case, for illustrative purposes, we use
        # x1, y1, x2, y2 order.
        boxes = [[0, 0, 10, 10], [10, 20, 50, 50]]
        # and labels
        labels = torch.tensor([10, 20])

        # create a BoxList from the boxes
        boxlist = BoxList(boxes, image.size, mode="xyxy")
        # add the labels to the boxlist
        boxlist.add_field("labels", labels)

        if self.transforms:
            image, boxlist = self.transforms(image, boxlist)

        # return the image, the boxlist and the idx in your dataset
        return image, boxlist, idx

    def get_img_info(self, idx):
        # get img_height and img_width. This is used if
        # we want to split the batches according to the aspect ratio
        # of the image, as it can be more efficient than loading the
        # image from disk
        return {"height": img_height, "width": img_width}
```

**Comment by fmassa** [comment]
@yxchng can you send a PR improving the README instructions on that part?

**Comment by yxchng** [comment]
@fmassa Sure, I will do it after I get the code to run with my own dataset.

**Comment by Wahaha1314** [comment]
> ## 🐛 Bug
> Very long waiting time before training starts (15 mins). I am not sure if it is a bug but it feels like one. If not, what do I need to do to speed it up?
> 
> > ...
> > 2019-04-09 13:58:35,545 maskrcnn_benchmark.data.build WARNING: When using more than one image per GPU you may encounter an out-of-memory (OOM) error if your GPU does not have sufficient memory. If this happens, you can reduce SOLVER.IMS_PER_BATCH (for training) or TEST.IMS_PER_BATCH (for inference). For training, you must also adjust the learning rate and schedule length according to the linear scaling rule. See for example: https://github.com/facebookresearch/Detectron/blob/master/configs/getting_started/tutorial_1gpu_e2e_faster_rcnn_R-50-FPN.yaml#L14
> > 2019-04-09 14:13:06,296 maskrcnn_benchmark.trainer INFO: Start training
> 
> * PyTorch Version (e.g., 1.0):
> * OS (e.g., Linux): 18.04
> * How you installed PyTorch (`conda`, `pip`, source): pip
> * Build command you used (if compiling from source):
> * Python version: 3.6.7
> * CUDA/cuDNN version: 10/7.4
> * GPU models and configuration: 1070 Ti
> * Any other relevant information:

您好，我想请问下，正式开始训练的时候确实要等待很长的时间吗？我在训练coco数据集时一直不动，也不报错。还有您说有没有高效的方法？
